### ISSUE TITLE
Proposal: submit for uiForm

### ISSUE THREAD

**Issue by emersion** [andlabs/libui#189]
Right now pressing the Enter key when typing in a `uiForm` does nothing. It should submit the form.

One solution would be to add a `uiFormOnSubmitted` function that calls the callback when Enter is pressed in the form's `uiEntry` childs. A `uiFormSubmit` function can be added to submit the form when clicking a button.

Thoughts?

**Comment by andlabs** [comment]
uiForm does not have a concept of submitting, because it is purely a layout container.

What you are describing would be done with the "default button" of a dialog. When I add custom dialogs, you will be able to do exactly this, and with any layout container.

**Comment by PaulBGD** [comment]
Custom dialogs sound awesome! You see them a lot in GTK and macOS. Do you have any sort of ETA on those?

**Comment by andlabs** [comment]
No, because dialogs are held up by OSs being highly incompatibly different with issues of modality. I don't know how to solve it, and explaining it will be more complex than explaining the menu issue.

**Comment by bcampbell** [comment]
I'm looking at some tangentially dialog-related issues at the moment, I'll add my notes here for future reference:

On Gtk, there's [GtkDialog](https://developer.gnome.org/gtk3/stable/GtkDialog.html):

- it has one upper area for adding custom widget layouts
- it has one lower area, a horizontal box for adding buttons which submit the dialog, each with  it's own user-defined ID (eg ok, cancel, replace, retry, yes, no, whatever)
- the dialog can be either modal or non-modal
- [GtkEntry](https://developer.gnome.org/gtk3/stable/GtkEntry.html) widgets in the upper area can be set to activate the default dialog button (the entry widget has a `activates-default` boolean property to set this behavour). Not sure about other widget types.

**Comment by bcampbell** [comment]
I think the [NSPanel](https://developer.apple.com/documentation/appkit/nspanel) window subclass is the OSX equivalent:

- [How Panels Work](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/WinPanel/Concepts/UsingPanels.html#//apple_ref/doc/uid/20000224)
- can be run modal or non-modal
- some controls (eg [NSTextField](https://developer.apple.com/documentation/appkit/nstextfield) look like they'll activate the panel's default action when enter is pressed. See [textDidEndEditing](https://developer.apple.com/documentation/appkit/nstextfield/1399420-textdidendediting).

**Comment by bcampbell** [comment]
Windows [Dialog Box reference](https://msdn.microsoft.com/en-us/library/windows/desktop/ms632588(v=vs.85).aspx)

- can be modal or non-modal ('modeless')
- seems really biased toward constucting the dialog layout using dialog templates from resource files. Ugh. Not clear to me yet if you can populate a dialog box layout programatically.
- Raymond Chen has what looks like a great multipart blog series on dialogs starting here: [The dialog manager, part 1: Warm-ups](https://blogs.msdn.microsoft.com/oldnewthing/20050329-00/?p=36043)

**Comment by bcampbell** [comment]
Looks like on windows you can, in fact, subclass the Dialog Box and create the layout yourself without faffing about with a dialog template.

- Example: http://zetcode.com/gui/winapi/dialogs/
- another Raymond Chen post on subclassing Dialogs: ['Another different type of dialog procedure'](https://blogs.msdn.microsoft.com/oldnewthing/20031113-00/?p=41843)

**Comment by andlabs** [comment]
GTK+: you do not have to follow the standard widget layout for dialog boxes. You don't even have to use GtkDialog as GtkWindow already has all the relevant properties in it, but I forget if there are any edge cases that make this not-ideal.

Windows: you do not have to use the dialog template system at all. In fact, you don't even have to use `WC_DIALOG`; Raymond Chen has a series of articles in which you recreate the functionality of the `DialogBox()` function!

The platform-incompatible issues I mentioned are much more subtle than anything you listed here. I can talk about them in more detail later.
